### PR TITLE
Optimized file format

### PR DESCRIPTION
**Pull request by ludvigarlebrink**
.asset file format handler added with .fbx to .asset converter. ONLY working for one model / fbx file.

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
Can one of the admins verify this patch?

File: src/Editor/GUI/Editors/ModelEditor.cpp [hunk 1]
```diff
@@ -47,7 +59,7 @@ void ModelEditor::SetVisible(bool visible) {
...
 void ModelEditor::FileSelected(const std::string& file) {
     model->extension = file.substr(file.find_last_of(".") + 1);
-    std::string destination = Hymn().GetPath() + FileSystem::DELIMITER + "Models" + FileSystem::DELIMITER + model->name + "." + model->extension;
-    FileSystem::Copy(file.c_str(), destination.c_str());
-    model->Load(file.c_str());
-}
+    destination = Hymn().GetPath() + FileSystem::DELIMITER + "Models" + FileSystem::DELIMITER + model->name + ".asset";
+    source = file;
+    hasSourceFile = true;
+}
```
Comment (Chainsawkitten) [inline src/Editor/GUI/Editors/ModelEditor.cpp]: Missing newline at end of file.

File: src/Editor/GUI/Editors/ModelEditor.hpp [hunk 2]
```diff
@@ -48,5 +48,13 @@ namespace GUI {
             FileSelector fileSelector;
             
             char name[128] = "";
+
+            bool hasSourceFile = false;
+            std::string source;
+            std::string destination;
+
+			bool triangulate = true;
```
Comment (Chainsawkitten) [inline src/Editor/GUI/Editors/ModelEditor.hpp]: Use spaces, not tabs.

File: src/Editor/Util/AssetConverter.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
+#include "AssetConverter.hpp"
+#include <Utility/Log.hpp>
+#include <Engine/Hymn.hpp>
+#include <Engine/Util/FileSystem.hpp>
+
+AssetConverter::AssetConverter()
+{
```
Comment (Chainsawkitten) [inline src/Editor/Util/AssetConverter.cpp]: Opening bracket on same line as function definition.

File: src/Editor/Util/AssetConverter.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	bool triangulate, bool importNormals, bool importTangents)
+{
+    Geometry::AssetFileHandler file;
+
+    // Return if file is not open.
+    if (!file.Open(destination, Geometry::AssetFileHandler::WRITE))
+        return false;
+	
+    unsigned int flags = triangulate ? aiProcess_Triangulate : 0;
+	flags = importNormals ? flags : flags | aiProcess_CalcTangentSpace;
```
Comment (Chainsawkitten) [inline src/Editor/Util/AssetConverter.cpp]: Use spaces, not tabs.

File: src/Editor/Util/AssetConverter.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    meshData->aabbDim = dim;
+    meshData->aabbOrigin = origin;
+    meshData->aabbMinpos = minValues;
+    meshData->aabbMaxpos = maxValues;
+
+    file->SaveStaticMesh(meshData);
+
+    delete meshData;
+    meshData = nullptr;
+}
```
Comment (Chainsawkitten) [inline src/Editor/Util/AssetConverter.cpp]: Missing newline at end of file.

File: src/Editor/Util/AssetConverter.hpp [hunk 6]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+#include <assimp/Importer.hpp>
+#include <assimp/scene.h>
+#include <assimp/postprocess.h>
+#include <Engine/Geometry/AssetFileHandler.hpp>
+#include <Video/Geometry/VertexType/StaticVertex.hpp>
+#include <Engine/Geometry/MathFunctions.hpp>
+
+class AssetConverter
```
Comment (Chainsawkitten) [inline src/Editor/Util/AssetConverter.hpp]: Document class.

File: src/Editor/Util/AssetConverter.hpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    	bool Convert(const char * filepath, const char * destination,
+    		bool triangulate, bool importNormals,
+    		bool importTangents);
+    
+    private:
+    	void ConvertMeshes(const aiScene * aScene, Geometry::AssetFileHandler * file);
+    	void ConvertMesh(aiMesh * mesh, Geometry::AssetFileHandler * file);
+    
+    	Assimp::Importer aImporter;
+};
```
Comment (Chainsawkitten) [inline src/Editor/Util/AssetConverter.hpp]: Missing newline at end of file.

File: src/Engine/Geometry/AssetFileHandler.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
+#include "AssetFileHandler.hpp"
+#include <Utility/Log.hpp>
+
+using namespace Geometry;
+
+AssetFileHandler::AssetFileHandler()
+{
```
Comment (Chainsawkitten) [inline src/Engine/Geometry/AssetFileHandler.cpp]: Opening bracket on same line as function definition.

File: src/Engine/Geometry/AssetFileHandler.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    wFile.write(reinterpret_cast<char*>(&numStaticMeshes), sizeof(uint16_t));
+}
+
+void AssetFileHandler::ClearMesh() {
+    if (staticMesh != nullptr) {
+        delete staticMesh;
+    }
+
+    staticMesh = nullptr;
+}
```
Comment (Chainsawkitten) [inline src/Engine/Geometry/AssetFileHandler.cpp]: Missing newline at end of file.

File: src/Engine/Geometry/AssetFileHandler.hpp [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * Start by using the Open() function. 
+     * The Open() function requries a filepath and
+     * a mode READ/WRITE.
+     * End by using the Close() function.
+     */
+    class AssetFileHandler {
+        public:
+            /// Mode to use class. READ/WRITE.
+        	enum Mode {
```
Comment (Chainsawkitten) [inline src/Engine/Geometry/AssetFileHandler.hpp]: Use spaces, not tabs.

File: src/Engine/Geometry/AssetFileHandler.hpp [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                }
+            };
+        
+            /// The current version of the exporter.
+        	const uint16_t CURRENT_VERSION = 1;
+        
+            /// Importing is supported from version.
+        	const uint16_t SUPPORTED_FROM = 1;
+        
+            /// Defualt constructor.
```
Comment (Chainsawkitten) [inline src/Engine/Geometry/AssetFileHandler.hpp]: Spelling error.

File: src/Engine/Geometry/AssetFileHandler.hpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        	AssetFileHandler(const char* filepath, Mode mode = READ);
+        
+            /// Destructor.
+        	~AssetFileHandler();
+        
+        	/// Open a .wkbf file.
+        	/**
+        	 * @param filepath Path of the file.
+        	 */
+        	bool Open(const char* filepath, Mode mode = READ);
```
Comment (Chainsawkitten) [inline src/Engine/Geometry/AssetFileHandler.hpp]: Document mode parameter.

File: src/Engine/Geometry/AssetFileHandler.hpp [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+            StaticMeshData * staticMesh = nullptr;
+        
+        	std::ifstream rFile;
+        	std::ofstream wFile;
+        	int fileVersion;
+        
+        	std::streampos globalHeaderStart;
+        	std::vector<std::streampos> staticMeshHeaderStart;
+    };
+}
```
Comment (Chainsawkitten) [inline src/Engine/Geometry/AssetFileHandler.hpp]: Missing newline at end of file.

File: src/Video/Geometry/Geometry3D.hpp [hunk 14]
```diff
@@ -67,9 +67,18 @@ namespace Video {
                 
                 /// Create local space axis-aligned bounding box around the geometry.
                 /**
-                 * @param positions Vector of vertex positions.
+                 * @param position Vector of vertex positions.
```
Comment (Chainsawkitten) [inline src/Video/Geometry/Geometry3D.hpp]: The parameter name is positions, not position.

File: src/Engine/Geometry/AssetFileHandler.cpp [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+bool AssetFileHandler::Open(const char* filepath, Mode mode) {
+    // Close the file is already open.
+    Close();
+
+    this->mode = mode;
+
+    if (mode == READ) {
+        // Open the .wkbf file
```
Comment (Chainsawkitten) [inline src/Engine/Geometry/AssetFileHandler.cpp]: Not .asset?

File: src/Engine/Geometry/AssetFileHandler.hpp [hunk 16]
```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <fstream>
+#include <cstdint>
+#include <assert.h>
+#include <vector>
+#include <Video/Geometry/VertexType/StaticVertex.hpp>
+
+namespace Geometry {
+    /// Handler for .res format.
```
Comment (Chainsawkitten) [inline src/Engine/Geometry/AssetFileHandler.hpp]: Not .asset?

**Pull request merged**